### ISSUE TITLE
Support Django 3.2

### ISSUE THREAD

**Issue by Jorl17** [jazzband/django-silk#465]
Currently, support for Django 3.2 is broken, as per ~~[this issue](https://github.com/jazzband/django-silk/issues/442).~~ (EDIT: the linked issue isn't related to Django 3.2, though support was broken before the PR I have submitted, and I now believe it is fixed)

Therefore, I created this issue to track progress on fixing django-silk in Django 3.2

**Comment by auvipy** [comment]
would you mind coming up with the initial PR?

**Comment by Jorl17** [comment]
Hi,

I wouldn't mind at all -- it was actually my original idea -- however, I'm a bit at a loss as to how to debug the issue. Do you have any slight guidelines as to try to find out where the problem is? I suppose somewhere some exception is being thrown and ignored, or maybe there's a specific bit of code that should be intercepting queries but isn't. Perhaps all I need to do to get a clearer idea is increase the log level?

If you could give me any pointers, as simple as they may be, I'd gladly try to fix the issue and contribute a PR.

Thanks!

**Comment by niksite** [comment]
I see a lot of exceptions like that:
```
ui_1                   |     silk_response = ResponseModelFactory(response).construct_response_model()
ui_1                   |   File "/app/lib64/python3.8/site-packages/silk/model_factory.py", line 302, in construct_response_model
ui_1                   |     raw_headers = self.response._headers
ui_1                   | AttributeError: 'Response' object has no attribute '_headers'
```

**Comment by Jorl17** [comment]
Interesting. I don't see any exceptions in my logs,but I'll look into it and try to setup a test project to debug this specific issue

**Comment by auvipy** [comment]
start with adding django 3.2 in build matrix and add django autofield settings in app.py checking the django 3.2 relese notes.

**Comment by dimi-tree** [comment]
🤔 I've installed from master:
```
pip install -e git+https://github.com/jazzband/django-silk.git#egg=silk
```
and I'm still seeing 0s for all queries (Django 3.2).

It'd be great to have support for Django 3.2 added into a new release when it is ready. 👍

**Comment by auvipy** [comment]
the issue was automatically closed.

**Comment by Jorl17** [comment]
I believe that Django 3.2 support is indeed fixed, but that issue #442 is something different, as shown by the tests and the test project. In my own project where we initially detected the problem with Django 3.2, installing from git with the applied fixes has solved our issues.

Notice that #442 involves showing 0 for all queries (in the case of Django 3.2 support, no value was shown at all).

See attached screenshot of test project working with Django 3.2
<img width="897" alt="Captura de ecrã 2021-04-13, às 10 43 29" src="https://user-images.githubusercontent.com/1668225/114533228-961b9a80-9c45-11eb-9985-7ca9340966f9.png">

@dimi-tree are you having the exact same issue as described in #442 (an explicit "0" for all queries) or are you seeing no numbers at all? Can you try to downgrade Django to 3.1 or 3.0 and check if the problem persists?

**Comment by dimi-tree** [comment]
@Jorl17 ah, yes, I think that's right - this is issue #442 where indeed there's an explicit "0" for all queries. Thank you for clarifying.

I've now downgraded to Django 3.1.8 and silk (v4.1.0) is working beautifully!

**Comment by yovelcohen** [comment]
Is there any hints as to what could cause the lib to not play nicely with 3.2?
downgrading is not always a possibility if you use a feature from 3.2

**Comment by davidhalter** [comment]
I have had the same "0 Queries" issue with Silk on Django 3.2. I haven't tested in a while, but I guess with Django 2.2 it still worked.

After taking a look at this, I can say that this happens only for me when I have multiple requests at the same time. So it looks like it has to do with Django/Silk's threading, but I'm not sure at all why it happens. I previously thought the queries don't get registered anymore and the mocking of the `SQLCompiler` was broken. However it turns out that it's probably something different.

**Comment by gevezex** [comment]
Any progress for django 3.2 support?  I am getting also these kinds of issues:

<img width="930" alt="Screenshot 2021-05-05 at 11 30 48" src="https://user-images.githubusercontent.com/4493063/117121813-6b7cb780-ad95-11eb-94ce-77acd59d6fbc.png">

**Comment by dmartin** [comment]
Are those with Django 3.2 issues using 4.1.0 or master? As https://github.com/jazzband/django-silk/commit/76a9b1e5dd173fce4216aad33a5dc7f9954c5f94 seems to have fixed the Django 3.2 issues for me, just waiting for a PyPI release.

**Comment by davidhalter** [comment]
I'm pretty sure I was using master and still had the issues with the `0 queries`. However that only appears if you start using simultaneous requests. Single requests don't reproduce this feature as far as I can tell.

**Comment by krzysieqq** [comment]
Any informations about when will be official release that supports Django 3.2?

**Comment by tiltec** [comment]
Can confirm that django-silk on `master` works with Django 3.2. PyPI release would be great!

**Comment by Archmonger** [comment]
@albertyw Are you available to create a PyPi release for the latest master build?

**Comment by albertyw** [comment]
I don't have the permissions to create a new PyPi release.  Since this is a jazzband project, you want one of the project leads (https://jazzband.co/projects/django-silk).  @avelis or @nasirhjafri .  It seems @nasirhjafri has been planning a release since #454 .

**Comment by Archmonger** [comment]
@albertyw seems like @nasirhjafri hasn't been on GH since May, and @avelis hasn't been since March. Is there anything that can be done to move things along regarding pypi releases?
Sorry for the pings about this, seems like you're the only (former?) maintainer for django-silk that's still active on GH at the moment.

**Comment by albertyw** [comment]
Since this is a jazzband project, the release process is documented in https://jazzband.co/about/releases.  I created https://github.com/jazzband/help/issues/233 and https://github.com/jazzband/help/issues/232 to get this done.  

As mentioned in https://github.com/jazzband/django-silk/issues/465#issuecomment-818504874, you can install django-silk directly from github, bypassing waiting for PyPI releases with

pip install -e git+https://github.com/jazzband/django-silk.git#egg=silk

Note 1: I recommend adding a `@<COMMIT_HASH>` after `django-silk.git` so your dependency installs are the same no matter what commits get merged into master.
Note 2: I recommend not installing from a git url if you don't have to, because git url installs are not cacheable.

**Comment by phoebebright** [comment]
Note: to install from repo need::

        pip install -e git+https://github.com/jazzband/django-silk.git#egg=django-silk

otherwise you get error::

      different name in metadata: 'django-silk'

**Comment by Archmonger** [comment]
> Note: to install from repo need::
> 
> ```
>     pip install -e git+https://github.com/jazzband/django-silk.git#egg=django-silk
> ```
> 
> otherwise you get error::
> 
> ```
>   different name in metadata: 'django-silk'
> ```
> 
> And not able to get it to run - seems like using django-silk not a good idea!

Correct, I have a PR open for this.

**Comment by hoanhan101** [comment]
If anyone is looking for a one-liner fix, here's what I put in my `requirements.txt` for `pip install`:
```
git+https://github.com/jazzband/django-silk@76a9b1e5dd173fce4216aad33a5dc7f9954c5f94#egg=django-silk
```

**Comment by tarikki** [comment]
Any updates on getting this to PyPi?

**Comment by Archmonger** [comment]
> Any updates on getting this to PyPi?

The project leads that would have been able to create releases for `django-silk` (@avelis and @nasirhjafri) are MIA.

Additionally, I've pinged the only remaining Jazzband roadie @auvipy (basically leader of the organization), but he is in bad health so his support is not available at the moment.

Which means that until  is better, django-silk will be deprecated.

**Comment by tiltec** [comment]
I wonder if we can get a note about missing Django 3.2 compatibility onto `README`? Possibly including the hint how to install from git. I just became Jazzband member, maybe I'll make that change...

**Comment by rohanahata** [comment]
@tiltec any updates on this?

**Comment by albertyw** [comment]
Still blocked on https://github.com/jazzband/help/issues/233.  Until someone is made a project lead for django-silk, or one of the original leads (@avelis or @nasirhjafri) becomes active again, this project is stuck.

**Comment by rohanahata** [comment]
@Archmonger can you weigh in on this?

**Comment by Archmonger** [comment]
@rohanahata Albert is correct. It's currently impossible to create a PyPi release due to Jazzband's organization structure.

Only two roles can make releases: Project leads and roadies.

Our project leads are MIA.

In this situation roadies are supposed to assign a new project lead and manually create a release. Unfortunately, there is only one roadie remaining, but he hasn't been receptive to communication, likely due to health issues.

We're completely unable to release until at least one of these individuals come back from the netherworld. Without roadie support, the whole organization has been freezing up situations like this for about a year 

I've been attempting to get in contact with the only remaining roadie @jezdez to see if I can become a roadie myself to move things forward. This situation is **extremely** problematic for the jazzband organization as a whole.

**Comment by Jorl17** [comment]
I opened this issue and have submitted the main patch that adds Django 3.2 support (although some work is still remaining, it seems). If there's anything I can do (I wouldn't mind experimenting with helping maintain the repo), please ring me up.

I'm all for helping and I'd be pleased to do so.

**Comment by Archmonger** [comment]
> I opened this issue and have submitted the main patch that adds Django 3.2 support (although some work is still remaining, it seems). If there's anything I can do (I wouldn't mind experimenting with helping maintain the repo), please ring me up.
> 
> I'm all for helping and I'd be pleased to do so.

Feel free to request my review when the PR is complete

**Comment by Archmonger** [comment]
@Jorl17 Just realized you might be asked "what is left to be 3.2 compatible".

To my knowledge, this is the only remaining 3.2 compatibility issue left on the main branch:
https://github.com/jazzband/django-silk/issues/489

**Comment by MarcoRizk** [comment]
Is this solved ?

**Comment by Archmonger** [comment]
I don't have any reason to believe there any 3.2 compatibility issues left on the main branch.

But unfortunately that's not in pypi.
